src/cose/sign1.rs

```diff
@@ -0,0 +1,484 @@
+            signature_payload,
+        })
+    }
+
+    /// Returns the signature payload that needs to be used to sign.
+    pub fn signature_payload(&self) -> &[u8] {
+        &self.signature_payload
+    }
+
+    /// Finalize the COSE_Sign1 by adding the signature.
+    pub fn finalize(self, signature: Vec<u8>) -> MaybeTagged<CoseSign1> {
+        let mut cose_sign1 = self.cose_sign1;
+        cose_sign1.inner.signature = signature;
+        cose_sign1
+    }
+}
+
+impl MaybeTagged<CoseSign1> {
+    /// Verify that the signature of a COSE_Sign1 is authentic.
+    pub fn verify<'a, V, S>(
+        &'a self,
+        verifier: &V,
+        detached_payload: Option<&[u8]>,
+        external_aad: Option<&[u8]>,
+    ) -> VerificationResult
+    where
+        V: Verifier<S> + SignatureAlgorithm,
+        S: TryFrom<&'a [u8]>,
+        S::Error: Into<signature::Error>,
+    {
+        if let Some(RegisteredLabelWithPrivate::Assigned(alg)) =
+            self.inner.protected.header.alg.as_ref()
+        {
+            if verifier.algorithm() != *alg {
+                return VerificationResult::Failure(
+                    "algorithm in protected headers did not match verifier's algorithm".into(),
+                );
+            }
+        }
+
+        let payload = match (self.inner.payload.as_ref(), detached_payload) {
+            (None, None) => return VerificationResult::Error(Error::NoPayload),
+            (Some(attached), None) => attached,
+            (None, Some(detached)) => detached,
+            _ => return VerificationResult::Error(Error::DoublePayload),
+        };
+
+        let signature = match S::try_from(self.inner.signature.as_ref())
+            .map_err(Into::into)
+            .map_err(Error::MalformedSignature)
+        {
+            Ok(sig) => sig,
+            Err(e) => return VerificationResult::Error(e),
+        };
+
+        let signature_payload = sig_structure_data(
+            SignatureContext::CoseSign1,
+            self.inner.protected.clone(),
+            None,
+            external_aad.unwrap_or_default(),
+            payload,
+        );
+
+        match verifier.verify(&signature_payload, &signature) {
+            Ok(()) => VerificationResult::Success,
+            Err(e) => VerificationResult::Failure(format!("signature is not authentic: {}", e)),
+        }
+    }
+
+    /// Retrieve the CWT claims set.
+    pub fn claims_set(&self) -> Result<Option<ClaimsSet>> {
+        match self.inner.payload.as_ref() {
+            None => Ok(None),
+            Some(payload) => ClaimsSet::from_slice(payload).map_or_else(
+                |e| Err(Error::UnableToDeserializeIntoClaimsSet(e)),
+                |c| Ok(Some(c)),
+            ),
+        }
+    }
+
+    /// If we are serialized as tagged.
+    pub fn is_tagged(&self) -> bool {
+        self.tagged
+    }
+
+    /// Set serialization to tagged.
+    pub fn set_tagged(&mut self) {
+        self.tagged = true;
+    }
```

Comment:
Not needed here if moved to `MaybeTagged<T>`

Reply:
moved